### PR TITLE
[Feature] Support DDOD: Disentangle Your Dense Object Detector(ACM MM2021 oral)

### PR DESCRIPTION
**Pull request by Irvingao**
# **Disentangle Your Dense Object Detector**

https://arxiv.org/pdf/2107.02963.pdf
## Introduction

Deep learning-based dense object detectors have achieved great success in the past few years and have been applied to numerous multimedia applications such as video understanding. However, the current training pipeline for dense detectors is compromised to lots of conjunctions that may not hold. In this paper, we investigate three such important conjunctions: 1) only samples assigned as positive in classification head are used to train the regression head; 2) classification and regression share the same input feature and computational fields defined by the parallel head architecture; and 3) samples distributed in different feature pyramid layers are treated equally when computing the loss. We first carry out a series of pilot experiments to show disentangling such conjunctions can lead to persistent performance improvement. Then, based on these findings, we propose Disentangled Dense Object Detector(DDOD), in which simple and effective disentanglement mechanisms are designed and integrated into the current state-of-the-art dense object detectors. Extensive experiments on MS COCO benchmark show that our approach can lead to 2.0 mAP, 2.4 mAP and 2.2 mAP absolute improvements on RetinaNet, FCOS, and ATSS baselines with negligible extra overhead. Notably, our best model reaches 55.0 mAP on the COCO test-dev set and 93.5 AP on the hard subset of WIDER FACE, achieving new state-of-the-art performance on these two competitive benchmarks.

## Results and Models

| Model | Backbone | Lr Schd | box mAP | AP50 | AP75 | APs | APm | APl |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |:---: |
| ATSS(IoU) | ResNet50 | 1x | 39.4 | 56.6 | 42.6 | 23.9 | 42.5 | 49.6 |
| DDOD | ResNet50 | 1x | 41.6 | 59.9 | 45.2 | 23.9 | 44.9 | 54.4 |
| DDOD-FCOS | ResNet50 | 1x | 41.6 | 59.9 | 45.3 | 24.0 | 44.6 | 54.8 |

## Modification

we aim to add the DDOD model for mmdet.

## Checklist

1. `configs/ddod/ddod_r50_1x_coco.py`
2. `configs/ddod/ddod_r50_1x_fcos_coco.py`
3. `mmdet/models/dense_heads/ddod_fcos_head.py`
4. `mmdet/models/dense_heads/ddod_head.py`

### CODE REVIEW

**Comment by hhaAndroid** [comment]
@Irvingao Thank you very much for your contribution. We will review the code soon.

**Comment by hhaAndroid** [comment]
@Irvingao Is the current accuracy all aligned?

File: configs/ddod/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+## Results and Models
+
+| Model | Backbone | Lr Schd | box mAP | AP50 | AP75 | APs | APm | APl |
```
Comment (hhaAndroid) [inline configs/ddod/README.md:13]: Please refer to https://github.com/open-mmlab/mmdetection/blob/master/configs/retinanet/README.md#results-and-models

**Comment by Irvingao** [comment]
> @Irvingao Is the current accuracy all aligned?

yes.

File: configs/ddod/ddod_r50_1x_coco.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
+fp16 = dict(loss_scale=512.)
```
Comment (hhaAndroid) [inline configs/ddod/ddod_r50_1x_coco.py:1]: If you need to use fp16, then the configuration file name should be changed to `ddod_r50_fp16_1x_coco.py`

Comment (hhaAndroid) [inline configs/ddod/ddod_r50_1x_coco.py:1]:
```python
_base_ = [
    '../_base_/datasets/coco_detection.py',
    '../_base_/schedules/schedule_1x.py', '../_base_/default_runtime.py'
]
```

File: configs/ddod/ddod_r50_1x_fcos_coco.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
+fp16 = dict(loss_scale=512.)
```
Comment (hhaAndroid) [inline configs/ddod/ddod_r50_1x_fcos_coco.py:1]: `ddod_fcos_r50_fp16_1x.py`

File: mmdet/core/bbox/assigners/atss_cost_assigner.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
+import torch
+
+from ..builder import BBOX_ASSIGNERS
+from ..iou_calculators import build_iou_calculator
+from .assign_result import AssignResult
+from .base_assigner import BaseAssigner
+
+def diou_loss(pred, target, eps=1e-7):
```
Comment (hhaAndroid) [inline mmdet/core/bbox/assigners/atss_cost_assigner.py:8]: Can you add `diou` calculation function to [bbox_overlaps](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/core/bbox/iou_calculators/iou2d_calculator.py#L75), it will be better to use

File: mmdet/core/bbox/assigners/atss_cost_assigner.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+                labelled as `ignored`, e.g., crowd boxes in COCO.
+            gt_labels (Tensor, optional): Label of gt_bboxes, shape (k, ).
+
+        Returns:
+            :obj:`AssignResult`: The assign result.
+        """
+        INF = 100000000
+        # NOTE first convert anchor to prediction bbox
+        bboxes = bboxes[:, :4]      # anchor bbox
+        bbox_preds = bbox_preds.detach()
```
Comment (hhaAndroid) [inline mmdet/core/bbox/assigners/atss_cost_assigner.py:126]: You can directly add `@torch.no_grad()` to this function

File: mmdet/models/dense_heads/ddod_head.py [hunk 6]
```diff
@@ -0,0 +1,796 @@
...
+                        reduce_mean, unmap)
+from ..builder import HEADS, build_loss
+from .anchor_head import AnchorHead
+from mmdet.core.bbox import bbox_overlaps
+
+EPS = 1e-12
+
+@HEADS.register_module()
+class DDODHead(AnchorHead):
+    """Bridging the Gap Between Anchor-based and Anchor-free Detection via
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/ddod_head.py:17]: Modify docs

File: mmdet/models/dense_heads/ddod_head.py [hunk 7]
```diff
@@ -0,0 +1,796 @@
...
+                        images_to_levels, multi_apply, multiclass_nms,
+                        reduce_mean, unmap)
+from ..builder import HEADS, build_loss
+from .anchor_head import AnchorHead
+from mmdet.core.bbox import bbox_overlaps
+
+EPS = 1e-12
+
+@HEADS.register_module()
+class DDODHead(AnchorHead):
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/ddod_head.py:16]: This head is similar to atss head, is it possible to inherit atss head？

**Comment by shinya7y** [comment]
The code should be cleaned before detailed review.
I believe the cleaning of the following commit is a good starting point.
https://github.com/shinya7y/UniverseNet/commit/173605e919c966c4d90a1030b71cc375023a90a8

File: mmdet/models/dense_heads/ddod_head.py (hunk 7 above)
Comment (shinya7y) [reply mmdet/models/dense_heads/ddod_head.py:16]: It will be better not to inherit ATSSHead because DDOD modifies various methods.

**Comment by shinya7y** [comment]
> > @Irvingao Is the current accuracy all aligned?
>
> yes.

The AP values in `configs/ddod/README.md` are the same as the [official README.md](https://github.com/zehuichen123/DDOD/blob/main/README.md).
What are the AP values of the models you trained?

File: mmdet/models/dense_heads/ddod_head.py (hunk 7 above)
Comment (hhaAndroid) [reply mmdet/models/dense_heads/ddod_head.py:16]: OK

**Comment by hhaAndroid** [comment]
@Irvingao Please develop based on the latest dev code. And retrain based on the code you submitted to see if the training and test accuracy are aligned？

File: configs/ddod/README.md (hunk 1 above)
Comment (Irvingao) [reply configs/ddod/README.md:13]: > @Irvingao Please develop based on the latest dev code. And retrain based on the code you submitted to see if the training and test accuracy are aligned？

Thanks for your reply. I will change to dev branch and resubmit the code after the model is retrained in the lastest version.

**Comment by Irvingao** [comment]
@hhaAndroid How can I update the model to `https://download.openmmlab.com/mmdetection/xxxx`and get the downloading url ?

**Comment by hhaAndroid** [comment]
> @hhaAndroid How can I update the model to `https://download.openmmlab.com/mmdetection/xxxx`and get the downloading url ?

You don't need to upload for now. We'll help upload it when the PR is done and then retrain.

**Comment by Irvingao** [comment]
@hhaAndroid I change branch from master to dev, slove all mentioned problems above and submit [a new PR of DDOD](https://github.com/open-mmlab/mmdetection/pull/7279).
